### ISSUE TITLE
Path-tracing with Nonlinear projections (for wider FOVs)

### ISSUE THREAD

**Issue by shaunlebron** [cschied/q2vkpt#10]
Congrats on the project 🎉 I wanted to add that path-tracing isn't just the holy grail for lighting—it also frees the renderer to use different projections for showing wider FOVs without terrible distortion:

![](https://github.com/shaunlebron/blinky/raw/master/readme-img/old-and-new.jpg)

- I explored this in [Quake](https://www.youtube.com/watch?v=f9v_XN7Wxh8), but only using the software renderer.
- We tried it in [Minecraft](https://www.youtube.com/watch?v=ImXwLXYBImA) by rendering a cubemap in OpenGL, but couldn't get better than 30 fps.
- It inspired partial inclusion in the [Unreal engine](https://docs.unrealengine.com/en-us/Engine/Rendering/PostProcessEffects/PaniniProjection), but with only standard FOVs to save frame rate.

Before I jump in—could you weigh in with your thoughts?  I imagine this wouldn't be hard to do since each pixel just has to map to a different ray determined by the chosen projection, but I'm not sure if anything in the pipeline assumes the standard projection.  Thanks for your thoughts.

**Comment by cschied** [comment]
This is entirely possible. Care needs to be taken that the forward-projection in [](https://github.com/cschied/q2vkpt/blob/master/src/refresh/vkpt/shader/asvgf_fwd_project.comp) is changed accordingly.

The main path tracer lives in [](https://github.com/cschied/q2vkpt/blob/master/src/refresh/vkpt/shader/path_tracer.h), search for `global_ubo.VP`

Some of the reprojection code and the reconstruction filter might need some fiddling with the epsilons.

**Comment by cschied** [comment]
Yes, you just need to make sure to also change all the code that uses MVP

**Comment by BurningNorth** [comment]
My practical knowledge of path tracing is too limited to be any help here, but as someone who's spent far too much time trying to come up with needlessly clever raster projections to do this in other engines without the post-process step (spoiler alert: haha nope), I'm super excited to see if you can make this work. 😃

**Comment by cschied** [comment]
This is the view-projection matrix, going from world-space to clip-space.

**Comment by cschied** [comment]
The plan looks good! Let me know if you need any additional information. There might be the need to fix some of the reconstruction filter but we'll have to see the results first.
